src/library/scala/concurrent/Future.scala

```diff
@@ -115,9 +115,7 @@ trait Future[+T] extends Awaitable[T] {
    */
   @deprecated("use `foreach` or `onComplete` instead (keep in mind that they take total rather than partial functions)", "2.12.0")
   def onSuccess[U](pf: PartialFunction[T, U])(implicit executor: ExecutionContext): Unit = onComplete {
-    case Success(v) =>
-      pf.applyOrElse[T, Any](v, Predef.identity[T]) // Exploiting the cached function to avoid MatchError
-    case _ =>
+    t => if (t.isInstanceOf[Success[T]]) pf.applyOrElse[T, Any](t.asInstanceOf[Success[T]].value, Future.id[T])
```

Comment:
I wonder if these kinds of changes are really relevant when we afterwards dispatch a runnable on a scheduler?

Reply:
Probably not, but I think benchmarks will have to tell whether deferring this decision is worth it.

Reply:
(I just pushed my latest code to this branch: https://github.com/viktorklang/scala/tree/wip-perf-futures-%E2%88%9A)